compiler/rustc_lint/src/non_local_def.rs

```diff
@@ -0,0 +1,201 @@
+use rustc_hir::{
+    def::DefKind, Body, FnRetTy, GenericArg, GenericParam, GenericParamKind, Item, ItemKind, Path,
+    QPath, TyKind,
+};
+use rustc_span::{def_id::DefId, sym, symbol::kw, MacroKind};
+
+use crate::{lints::NonLocalDefinitionsDiag, LateContext, LateLintPass, LintContext};
+
+declare_lint! {
+    /// The `non_local_definitions` lint checks for `impl` blocks and `#[macro_export]`
+    /// macro inside bodies (functions, enum discriminant, ...).
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// trait MyTrait {}
+    /// struct MyStruct;
+    ///
+    /// fn foo() {
+    ///     impl MyTrait for MyStruct {}
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Creating non local definitions go against expectation and can create decrepencies
+    /// in tooling. In should be avoided.
```

Comment:
Would be a good idea to add the precise (to some level at least) description of how the check works. I.e. explain that implementing "local" traits or for "local" types is ok. Similarly mention that impls inside `const _` are also ok (potentially mentioning that it's an open question).

Reply:
Done. Let me know if I should improve further.